executor/aggregate.go

```diff
@@ -45,39 +96,482 @@ type HashAggExec struct {
+			aggCtxsMap:          make(aggCtxsMapper, 0),
+			groupSet:            mvmap.NewMVMap(),
+			inputCh:             e.partialOutputChs[i],
+			outputCh:            e.finalOutputCh,
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+// HashAggPartialWorker gets and handles origin data or partial data from inputCh,
+// then shuffle the intermediate results to corresponded final workers.
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup, finalConcurrency int) {
+	needShuffle, sc := false, ctx.GetSessionVars().StmtCtx
+	defer func() {
+		if needShuffle {
+			w.shuffleIntermData(sc, finalConcurrency)
+		}
+		waitGroup.Done()
+	}()
+	for {
+		select {
+		case <-w.finishCh:
+			return
+		case chk, ok := <-w.inputCh:
+			if !ok {
+				return
+			}
+			w.chk.SwapColumns(chk)
+			w.giveBackCh <- &HashAggInput{
+				chk:        chk,
+				giveBackCh: w.inputCh,
+			}
+		}
+
+		if err := w.updateIntermData(ctx, sc, w.chk, len(w.aggCtxsMap)); err != nil {
+			w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return
+		}
+		needShuffle = true
+	}
+}
+
+func (w *HashAggPartialWorker) updateIntermData(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (err error) {
+	inputIter := chunk.NewIterator4Chunk(chk)
+	for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
+		groupKey, err := w.getGroupKey(sc, row)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		aggEvalCtxs := w.getContext(sc, groupKey, w.aggCtxsMap)
+		for i, af := range w.aggFuncs {
+			if af.Update(aggEvalCtxs[i], sc, row) != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
 	return nil
 }
 
+// shuffleIntermData shuffles the intermediate data of partial workers to corresponded final workers.
+// We only support parallel execution for single-machine, so process of encode and decode can be skipped.
+func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, finalConcurrency int) {
+	groupKeyChks := make([]*chunk.Chunk, finalConcurrency)
+	for groupKey := range w.aggCtxsMap {
+		finalWorkerIdx := int(murmur3.Sum32([]byte(groupKey))) % finalConcurrency
+		if groupKeyChks[finalWorkerIdx] == nil {
+			groupKeyChks[finalWorkerIdx] = chunk.NewChunkWithCapacity([]*types.FieldType{types.NewFieldType(mysql.TypeVarString)}, len(w.aggCtxsMap)/finalConcurrency)
+		}
+		groupKeyChks[finalWorkerIdx].AppendString(0, groupKey)
```

Comment:
I still argue that we don't need to use `Chunk` to store group keys: this may introduce many memory copies on these group keys. If we use a `[][]byte`, we just need to append these group keys to a slice, and no more memory consumption and memory copy overhead. This is helpful especially when we have a lot of group keys.